core/provider/digitalocean/provider.go

```diff
@@ -0,0 +1,407 @@
+}
+
+// NewProviderWithClient creates a provider with custom digitalocean/docker client implementation.
+// This is primarily used for testing.
+func NewProviderWithClient(ctx context.Context, logger *zap.Logger, providerName string, doClient DoClient, dockerClients map[string]DockerClient, additionalUserIPS []string, sshKeyPair *SSHKeyPair) (*Provider, error) {
+	if sshKeyPair == nil {
+		newSshKeyPair, err := MakeSSHKeyPair()
+		if err != nil {
+			return nil, err
+		}
+		sshKeyPair = newSshKeyPair
+	}
+
+	userIPs, err := getUserIPs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	userIPs = append(userIPs, additionalUserIPS...)
+
+	if dockerClients == nil {
+		dockerClients = make(map[string]DockerClient)
+	}
+
+	digitalOceanProvider := &Provider{
+		logger:        logger.Named("digitalocean_provider"),
+		doClient:      doClient,
+		dockerClients: dockerClients,
+		state: &ProviderState{
+			TaskStates: make(map[int]*TaskState),
+			UserIPs:    userIPs,
+			Name:       providerName,
+			SSHKeyPair: sshKeyPair,
+			PetriTag:   fmt.Sprintf("petri-droplet-%s", util.RandomString(5)),
+		},
+	}
+
+	_, err = digitalOceanProvider.createTag(ctx, digitalOceanProvider.state.PetriTag)
+	if err != nil {
+		return nil, err
+	}
+
+	firewall, err := digitalOceanProvider.createFirewall(ctx, userIPs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create firewall: %w", err)
+	}
+
+	digitalOceanProvider.state.FirewallID = firewall.ID
+
+	//TODO(Zygimantass): TOCTOU issue
+	if key, _, err := doClient.GetKeyByFingerprint(ctx, sshKeyPair.Fingerprint); err != nil || key == nil {
+		_, err = digitalOceanProvider.createSSHKey(ctx, sshKeyPair.PublicKey)
+		if err != nil {
+			if !strings.Contains(err.Error(), "422") {
+				return nil, err
+			}
+		}
+	}
+
+	return digitalOceanProvider, nil
+}
+
+func (p *Provider) CreateTask(ctx context.Context, definition provider.TaskDefinition) (provider.TaskI, error) {
+	if err := definition.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("failed to validate task definition: %w", err)
+	}
+
+	if definition.ProviderSpecificConfig == nil {
+		return nil, fmt.Errorf("digitalocean specific config is nil for %s", definition.Name)
+	}
+
+	var doConfig DigitalOceanTaskConfig
+	doConfig, ok := definition.ProviderSpecificConfig.(DigitalOceanTaskConfig)
+	if !ok {
+		return nil, fmt.Errorf("invalid provider specific config type for %s", definition.Name)
+	}
+
+	if err := doConfig.ValidateBasic(); err != nil {
+		return nil, fmt.Errorf("could not cast digitalocean specific config: %w", err)
+	}
+
+	p.logger.Info("creating droplet", zap.String("name", definition.Name))
+
+	droplet, err := p.CreateDroplet(ctx, definition)
+	if err != nil {
+		return nil, err
+	}
+
+	ip, err := droplet.PublicIPv4()
+	if err != nil {
+		return nil, err
+	}
+
+	p.logger.Info("droplet created", zap.String("name", droplet.Name), zap.String("ip", ip))
+
+	if p.dockerClients[ip] == nil {
```

Comment:
cache docker client here, so you don't have to reference p.dockerClients[ip] every time